src/server/process_manager.rs

```diff
@@ -2,83 +2,427 @@
+        logger: &Logger,
+    ) -> Result<(), KillingChildError> {
+        match self.children.get_mut(name) {
+            // if the given program have running process we iterate over them and send to them the correct signal
+            Some(processes) => {
+                /*
+                we need to decide if we keep the child around for this we need to know
+                if the process must be killed or gracefully shutdown
+                if gracefully shutdown we need to keep the process else if the kill
+                happened correctly we need to get ride of them if the kill didn't happened correctly
+                we need to exit and warn the user that this did'nt go according to plan. we can't however
+                remove the process
+                */
+                match config.programs.get(name) {
+                    Some(config) => {
+                        // here we keep the process just sending them the signal required by the config
+                        processes.iter_mut().for_each(|process| {
+                            log_info!(
+                                logger,
+                                "about to gracefully shutdown process number : {}",
+                                process.get_child_id()
+                            );
+                            process.send_signal(&config.stop_signal);
+                        });
+                        Ok(())
+                    }
+                    None => {
+                        // here we killed them
+                        // PS if you want to learn more about closure you can try to transform the for loop in the iter version to see what happen
+                        for process in processes.iter_mut() {
+                            log_info!(
+                                logger,
+                                "about to kill process number : {}",
+                                process.get_child_id()
+                            );
+                            process.kill()?;
+                        }
+                        Ok(())
+                    }
+                }
+            }
+            // if no process are found for a given name we return the corresponding error
+            None => {
+                log_info!(
+                    logger,
+                    "tried to remove child of process: {name} but none where found"
+                );
+                Err(KillingChildError::NoProgramFound)
+            }
+        }
+        // here we don't remove the entry since the monitor will do it for use
     }
 
     /// do one round of monitoring
-    fn monitor_once(&mut self, shared_config: SharedConfig) {
-        // check the status of all the child
+    fn monitor_once(&mut self, config: &RwLock<Config>, logger: &Logger) {
```

Comment:
I think it's possible to separate the 4 iteration blocks into separate sub-functions